# Assignment 1
### Understanding Uncertainty
### Due 9/5

1. Create a new public repo on Github under your account. Include a readme file.
2. Clone it to your machine. Put this file into that repo.
3. Use the following function to download the example data for the course:

In [1]:
def download_data(force=False):
    """Download and extract course data from Zenodo."""
    import urllib.request, zipfile, os
    
    zip_path = 'data.zip'
    data_dir = 'data'
    
    if not os.path.exists(zip_path) or force:
        print("Downloading course data")
        urllib.request.urlretrieve(
            'https://zenodo.org/records/16954427/files/data.zip?download=1',
            zip_path
        )
        print("Download complete")
    else:
        print("Download file already exists")
        
    if not os.path.exists(data_dir) or force:
        print("Extracting data files...")
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            zip_ref.extractall(data_dir)
        print("Data extracted")
    else:
        print("Data directory already exists")

download_data()

Download complete
Extracting data files...
Data extracted


4. Open one of the datasets using Pandas:
    1. `ames_prices.csv`: Housing characteristics and prices
    2. `college_completion.csv`: Public, nonprofit, and for-profit educational institutions, graduation rates, and financial aid
    3. `ForeignGifts_edu.csv`: Monetary and in-kind transfers from foreign entities to U.S. educational institutions
    4. `iowa.csv`: Liquor sales in Iowa, at the transaction level
    5. `metabric.csv`: Cancer patient and outcome data
    6. `mn_police_use_of_force.csv`: Records of physical altercations between Minnessota police and private citizens
    7. `nhanes_data_17_18.csv`: National Health and Nutrition Examination Survey
    8. `tuna.csv`: Yellowfin Tuna Genome (I don't recommend this one; it's just a sequence of G, C, A, T )
    9. `va_procurement.csv`: Public spending by the state of Virginia

In [26]:
import pandas as pd
gifts = pd.read_csv("data/ForeignGifts_edu.csv")
gifts.head()

ID   OPEID               Institution Name          City State  \
0   1  102000  Jacksonville State University  Jacksonville    AL   
1   2  104700                Troy University          Troy    AL   
2   3  105100          University of Alabama    Tuscaloosa    AL   
3   4  105100          University of Alabama    Tuscaloosa    AL   
4   5  105100          University of Alabama    Tuscaloosa    AL   

   Foreign Gift Received Date  Foreign Gift Amount      Gift Type  \
0                       43738               250000  Monetary Gift   
1                       43592               463657       Contract   
2                       43466              3649107       Contract   
3                       43472                 1000       Contract   
4                       43479                49476       Contract   

  Country of Giftor                          Giftor Name  
0             CHINA                                  NaN  
1             CHINA     Confucius Institute Headquarters  
2           ENGLAND  Springer Nature Customer Service Ce  
3      SAUDI ARABIA       Saudi Arabia Education Mission  
4      SAUDI ARABIA       Saudi Arabia Education Mission

5. Pick two or three variables and briefly analyze them
    - Is it a categorical or numeric variable?
    - How many missing values are there? (`df['var'].isna()` and `np.sum()`)
    - If categorical, tabulate the values (`df['var'].value_counts()`) and if numeric, get a summary (`df['var'].describe()`)

The variables I'm choosing are "City", "Foreign Gift Amount", and "Country of Giftor". 

"City" and "Country of Giftor" are categorical. "Foreign Gift Amount" is numeric. 

In [20]:
print(gifts['City'].isna().sum())
print(gifts['Foreign Gift Amount'].isna().sum())
print(gifts['Country of Giftor'].isna().sum())

print(gifts['Giftor Name'].isna().sum())

0
0
0
3751


"City", "Foreign Gift Amount", and "Country of Giftor" all have zero missing values. I was curious about this, so I also found the number of missing values in "Giftor Name". It has 3751, meaning not all the columns are perfectly filled in. 

In [ ]:
pd.set_option('display.max_rows', None) # added this line to be able to see more of the values
gifts['City'].value_counts()

City
Los Angeles           4218
Pasadena              2502
New York              1507
Columbus              1238
Baltimore             1084
Cambridge             1038
Ann Arbor              715
Philadelphia           703
Stanford               668
Ithaca                 653
New Haven              550
Chicago                523
Pittsburgh             469
Denver                 439
Tempe                  403
Berkeley               394
Fort Collins           379
Austin                 377
College Station        376
Boulder                363
Babson Park            318
Glenside               304
Houston                299
Jamaica                288
Chapel Hill            286
Atlanta                264
Washington             258
Louisville             253
Evanston               251
Seattle                245
Indianapolis           244
Birmingham             241
Princeton              219
Boston                 218
La Jolla               203
Cleveland              189
Champaign              

In [29]:
gifts['Country of Giftor'].value_counts()

Country of Giftor
ENGLAND                     3655
CHINA                       2461
CANADA                      2344
JAPAN                       1896
SWITZERLAND                 1676
SAUDI ARABIA                1610
FRANCE                      1437
GERMANY                     1394
HONG KONG                   1080
SOUTH KOREA                  811
QATAR                        693
THE NETHERLANDS              512
KOREA                        452
INDIA                        434
TAIWAN                       381
ISRAEL                       373
UNITED ARAB EMIRATES         365
SINGAPORE                    360
AUSTRALIA                    335
ITALY                        318
KUWAIT                       302
DENMARK                      292
SWEDEN                       275
BRAZIL                       256
SPAIN                        240
NORWAY                       240
IRELAND                      224
MEXICO                       221
INDONESIA                    184
NIGERIA                  

In [30]:
gifts['Foreign Gift Amount'].describe()

count    2.822100e+04
mean     5.882327e+05
std      3.222011e+06
min     -5.377700e+05
25%      5.700000e+03
50%      9.461500e+04
75%      3.761420e+05
max      1.000000e+08
Name: Foreign Gift Amount, dtype: float64

6. What are some questions and prediction tools you could create using these data? Who would the stakeholder be for that prediction tool? What practical or ethical questions would it create? What other data would you want, that are not available in your data?

Using the country, the amount gifted, and even who gave it (in case that information provides relavent context), one potential tool would be predicting whether a gift is likely to be monetary or a contract, since those types of 'gifts' often come with very different stipulations.

If the stakeholder was a university, they could use the data to figure out which countries and which organizations within those countries to focus on to try and gain future fundraising or partnerships. Aa model could also predict the expected size of a gift from certain countries or donors, or even maybe flag gifts that might raise compliance or ethical concerns (ex: strangely large contracts from politically sensitive regions). I think possible stakeholders could also include the Department of Education, policymakers, and students and faculty who are indirectly impacted how these financial resources are used. 

Some ethical concerns would be about transparency about where a gift comes from or the expectations joining it, foreign influence, whether some institutions gain some sort of advantage through foreign influence, and if the gift is indicative of a problematic relationship. 

I think to strengthen these prediction models and make sure ethical concerns are considered, I'd want more detail on the purpose of the gift, which department or program the gift is meant to benefit (if there are specific recievers intended), and if there are any conditions attached to the gift. 

7. Commit your work to the repo (`git commit -am 'Finish assignment'` at the command line, or use the Git panel in VS Code). Push your work back to Github and submit the link on Canvas in the assignment tab.